sqlglot/lineage.py

```diff
@@ -94,29 +94,51 @@ def to_node(
                 )
             return node
 
-        select = next(select for select in scope.selects if select.alias_or_name == column_name)
+        # Find the specific select clause that is the source of the column we want.
+        # This can either be a specific, named select or a generic `*` clause.
+        select = None
+        for maybe_select in scope.selects:
+            if maybe_select.alias_or_name == column_name:
+                # If we get an exact match, stop searching.
+                select = maybe_select
+                break
+            elif isinstance(maybe_select, exp.Star):
```

Comment:
the elif is unncessary, just else is fine

Reply:
Hrmm, how so? Only a `*` or an exact match should be chosen - non matches should be ignored, no?

Reply:
what about this?

select = next(select for select in scope.selects if select.alias_or_name == column_name, exp.Star() if scope.expression.is_star else raise)

Reply:
@Sharadh since you do a break, you don't need an elif, just an if, but i think i have a more elegant solution

Reply:
Oh yeah, you meant:
> the elif is unncessary, just ~else~ if is fine

That makes sense!

> ```
> select = next(select for select in scope.selects if select.alias_or_name == column_name, exp.Star() if scope.expression.is_star else raise)
> ```
This wouldn't work if the `exp.Star()` appears first in the selects list; we'd want to check if a more specific select matches if both a `*` and a more specific select are found.

Reply:
the star trigger only happens when we can't find the explicit match right?

Reply:
So, consider this (the second element of the generator):
```
exp.Star() if scope.expression.is_star
```
That's only triggered if the expression is `SELECT * FROM foo`. But what about something like `SELECT *, x FROM foo`?
1. If we're searching for "y", we want the Star condition to be triggered
2. If we're searching for "x", we _don't_ want the Star condition to be triggered (we want the more specific x)

I think both of these cases ^ will not work under the proposed change?

Reply:
Ah, I wonder if the nuance I'm missing is: the optimizer would convert:
```
SELECT *, x FROM foo
```
to
```
SELECT y FROM foo
```
or
```
SELECT x FROM foo
```
for 1/ and 2/ respectively?

Reply:
doesn’t my solution provide that ? i’ll try it later

Reply:
```python
        select = next(
            (select for select in scope.selects if select.alias_or_name == column_name),
            exp.Star() if scope.expression.is_star else None,
        )

        if not select:
            raise ValueError(f"Could not find {column_name} in {scope.expression}")
```
i just did this and the tests pass, so i'll commit it

Reply:
Ack - I just realized `is_star` is overridden for `Select`: https://github.com/tobymao/sqlglot/blob/8dc4d802502045c635af1019eb94d8197be49b5c/sqlglot/expressions.py#L2676-L2678

This is indeed more elegant - thanks for the refactor!